### PR TITLE
Ports: Flying over tables in zero gravity with a jetpack

### PR DESCRIPTION
**Pull request by Super3222**
I decided to allow people to pass over tables when they have a jetpack ON, on their BACK, and a gravity check fails.

https://github.com/yogstation13/yogstation-old/pull/266
https://github.com/tgstation/tgstation/pull/15866

:cl: Super3222
tweak: You can now fly over tables with a jetpack when gravity is down!
/:cl:

### CODE REVIEW

**Comment by ParagonPrime** [comment]
I'd be partial to having zero G just flat out negate tables, but maybe that's too far.

File: code/game/objects/items/weapons/tanks/jetpack.dm [hunk 1]
```diff
@@ -35,23 +35,33 @@
...
-/obj/item/weapon/tank/jetpack/proc/turn_on()
+/obj/item/weapon/tank/jetpack/proc/turn_on(mob/user)
 	on = TRUE
+	var/mob/living/C = user
+
+	if(C.pass_flags & JETPACKTABLE) // a check to make sure we don't start with it again.
+		C.pass_flags &= ~JETPACKTABLE
+
+	if (!has_gravity(user))
+		C.pass_flags |= JETPACKTABLE
```
Comment (X-TheDark) [inline code/game/objects/items/weapons/tanks/jetpack.dm]: Won't this just check for gravity ONCE, when the jetpack is turned on, meaning that even if gravity comes on later, you still retain the tablepassing. Would you mind testing this with gravity being turned on AFTER you turn on the jetpack?

After you report the results of your testing (if it turns out true), I'll give you some things I think you need to move around/change.

Comment (Super3222) [inline code/game/objects/items/weapons/tanks/jetpack.dm]: I see what you're getting at...
I gave it a test, and it seems like you're right. So after turning gravity off and then back on again (leaving my jetpack toggled on)... It doesn't check for gravity, thus it still the player still carries `JETPACKTABLE` flag which allows you to still fly over tables.

Comment (X-TheDark) [inline code/game/objects/items/weapons/tanks/jetpack.dm]: To fix this, move the `!has_gravity(user)` check to somewhere after line 136 in tables_racks.dm (the new if you made). Add it as another && like `if(jetpacktable && jetpacktable.on && !has_gravity(C))`. Leave this as just `C.pass_flags |= JETPACKTABLE` .

Also, pretty sure the if in lines 52-53 is unnecessary. Why unset the flag if we are just going to set it anyway using bitflag operations?

**Comment by X-TheDark** [comment]
Also, what about dropping the active jetpack? Is there a check/something to turn the jetpack off/remove the bitflag when `dropped()` is called?

**Comment by GrayRachnid** [comment]
Mining cyborgs also have jetpacks

**Comment by Super3222** [comment]
> Also, what about dropping the active jetpack? Is there a check/something to turn the jetpack off/remove the bitflag when dropped() is called?

No, not currently. I made a check in tables_racks.dm right after the check for JETPACKTABLEwhich makes sure you have a jetpack on and it's toggled on.

**Comment by ParagonPrime** [comment]
Set the mineborg jetpack to be a special type that doesn't have this flag.

**Comment by Super3222** [comment]
> Also, what about dropping the active jetpack? Is there a check/something to turn the jetpack off/remove the bitflag when dropped() is called?

No, not currently, but I'll add a check eventually.
The reason for that is, in tables_racks.dm, it'll only allow you over the table if you specifically have a jetpack which is already turned on.

The check on 52-53 is so that someone doesn't turn on their jetpack, take it off, put on another jetpack, turn that one on, and then break themselves (where they'll be able to walk through airlocks and a lot of other broken shit).

**Comment by X-TheDark** [comment]
Wait, how does that issue even occur? I thought bitflag operation `|=` in BYOND just set the flag (if it is unset) or not do anything if the flag is already set. It's not the same as `+=`, which is the one that does what you described.

**Comment by Super3222** [comment]
Ah, right! I was referring to `+=`. My fault. It shouldn't happen

**Comment by Super3222** [comment]
@GrayRachnid @ParagonPrime 

All borgs have jetpacks. They just need an ion thruster module from robotics.
However, they won't be able to jump tables.

File: code/game/objects/items/weapons/tanks/jetpack.dm [hunk 2]
```diff
@@ -35,23 +35,26 @@
...
 
-/obj/item/weapon/tank/jetpack/proc/turn_on()
+/obj/item/weapon/tank/jetpack/proc/turn_on(mob/user)
 	on = TRUE
+	user.pass_flags |= JETPACKTABLE
 	icon_state = "[initial(icon_state)]-on"
 	ion_trail.start()
 
-/obj/item/weapon/tank/jetpack/proc/turn_off()
+/obj/item/weapon/tank/jetpack/proc/turn_off(mob/user)
```
Comment (oisin100) [inline code/game/objects/items/weapons/tanks/jetpack.dm]: Just wondering, Is this called when a player takes the jetpack off? If its not then bad things are going to happen.

Comment (Super3222) [inline code/game/objects/items/weapons/tanks/jetpack.dm]: I'm going to rewrite this, and not use bitflags. I've thought up an easier method.

Comment (oisin100) [inline code/game/objects/items/weapons/tanks/jetpack.dm]: Looked back at the code, I realized you had a check to see if they had a jetpack. My bad.

File: code/game/objects/items/weapons/tanks/jetpack.dm [hunk 3]
```diff
@@ -35,22 +35,22 @@
...
-		turn_off()
+		turn_off(user)
 		user << "<span class='notice'>You turn the jetpack off.</span>"
 	for(var/X in actions)
 		var/datum/action/A = X
 		A.UpdateButtonIcon()
 
 
-/obj/item/weapon/tank/jetpack/proc/turn_on()
+/obj/item/weapon/tank/jetpack/proc/turn_on(mob/user)
```
Comment (oisin100) [inline code/game/objects/items/weapons/tanks/jetpack.dm]: Not much point in the mob/user var anymore

Comment (Super3222) [inline code/game/objects/items/weapons/tanks/jetpack.dm]: Right

**Pull request merged**